flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/CountTriggerTest.java

```diff
@@ -0,0 +1,123 @@
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.FIRE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(2, 4)));
+
+		// right now, CountTrigger will clear it's state in onElement when firing
+		// ideally, this should be moved to onFire()
+		assertEquals(1, testHarness.numStateEntries());
+		assertEquals(0, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+	}
+
+	/**
+	 * Verify that clear() does not leak across windows.
+	 */
+	@Test
+	public void testClear() throws Exception {
+		TriggerTestHarness<Object, TimeWindow> testHarness =
+				new TriggerTestHarness<>(CountTrigger.<TimeWindow>of(3), new TimeWindow.Serializer());
+
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(2, 4)));
+
+		// shouldn't have any timers
+		assertEquals(0, testHarness.numProcessingTimeTimers());
+		assertEquals(0, testHarness.numEventTimeTimers());
+
+		assertEquals(2, testHarness.numStateEntries());
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+
+		testHarness.clearTriggerState(new TimeWindow(2, 4));
+
+		assertEquals(1, testHarness.numStateEntries());
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(0, testHarness.numStateEntries(new TimeWindow(2, 4)));
+
+		testHarness.clearTriggerState(new TimeWindow(0, 2));
+
+		assertEquals(0, testHarness.numStateEntries());
+		assertEquals(0, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(0, testHarness.numStateEntries(new TimeWindow(2, 4)));
+	}
+
+	@Test
+	public void testMergingWindows() throws Exception {
+		TriggerTestHarness<Object, TimeWindow> testHarness =
+				new TriggerTestHarness<>(CountTrigger.<TimeWindow>of(3), new TimeWindow.Serializer());
+
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(0, 2)));
+		assertEquals(TriggerResult.CONTINUE, testHarness.processElement(new StreamRecord<Object>(1), new TimeWindow(2, 4)));
+
+		// shouldn't have any timers
+		assertEquals(0, testHarness.numProcessingTimeTimers());
+		assertEquals(0, testHarness.numEventTimeTimers());
+
+		assertEquals(2, testHarness.numStateEntries());
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(0, 2)));
+		assertEquals(1, testHarness.numStateEntries(new TimeWindow(2, 4)));
+
```

Comment:
You could test this a little bit more, e.g. with a wider window that also subsumes both, or 3 windows and just merging two, etc. to catch corner cases